src/bmeg/__init__.py

```diff
@@ -0,0 +1,265 @@
+
+    def schema(self):
+        return self._schema
+
+    def validate(self):
+        validate(self.props(), self.schema())
+        return
+
+    def label(self):
+        return self.__class__.__name__
+
+    def gid(self):
+        if not isinstance(self.id, self._gid_cls):
+            raise TypeError("expected id of type {}", self._gid_cls)
+        return self.id
+
+    def __repr__(self):
+        return '<%s(%s)>' % (self.__class__.__name__,
+                             self.props(preserve_null=True))
+
+    def __setattr__(self, key, item):
+        if key not in self._props:
+            raise KeyError("object does not contain key '{}'".format(key))
+        self._props[key] = item
+
+    def __getattr__(self, key):
+        return self._props[key]
+
+    def __getitem__(self, k):
+        return self.__getattr__(k)
+
+    def __setitem__(self, k, v):
+        return self.__setattr__(k, v)
+
+
+def capitalize(label):
+    label = "".join([(lambda x: x[0].upper() + x[1:])(x) for x in re.split("_| +", label)])
+    if label.startswith("G2p"):
+        label = "G2P" + label[3:]
+    return label
+
+
+class Vertex:
+    pass
+
+
+class Edge:
+    pass
+
+
+@enforce_types
+@dataclasses.dataclass(frozen=True)
+class Deadletter(Vertex):
+    """ standard way to log missing data """
+    target_label: str  # desired vertex label
+    data: dict  # data from source
+
+    def validate(self):
+        return
+
+    def label(self):
+        return self.__class__.__name__
+
+    def gid(self):
+        return Deadletter.make_gid(self.target_label, self.data)
+
+    def props(self):
+        return self.__dict__
+
+    @classmethod
+    def make_gid(cls, target_label, data):
+        # create hash from data dict
+        data = '%s:%s' % (target_label, str(data))
+        datahash = hashlib.sha1()
+        datahash.update(data.encode('utf-8'))
+        datahash = datahash.hexdigest()
+        return "%s:%s" % (cls.__name__, datahash)
+
+
+_schemaPath = pkg_resources.resource_filename(__name__, "bmeg-dictionary/gdcdictionary/schemas")
+_schema = BMEGDataDictionary(root_dir=_schemaPath)
+
+__all__ = ['Vertex', 'Edge', 'Deadletter']
+for k, schema in _schema.schema.items():
+    name = capitalize(schema["id"])
+    if name not in gid_factories:
+        print("using default gid factory for '{}'".format(name), file=sys.stderr)
+        gid_factory = partial(default_gid, name)
+    else:
+        gid_factory = gid_factories[name]
+    cls = type(
+        name,
+        (ClassInstance, Vertex),
+        {
+            '_schema': schema,
+            '_gid_cls': types.new_class("{}GID".format(name), (str,), {}),
+            'make_gid': classmethod(cast_gid(gid_factory))
+        }
+    )
+    globals()[name] = cls
+    __all__.append(name)
+
+for k, schema in _schema.schema.items():
+    for link in schema['links']:
+        # TODO: handle link subgroup?
```

Comment:
@bwalsh did you mention that you had code that demonstrates how to handle link subgroups?